diff --git a/mysql-test/r/partition_locking.result b/mysql-test/r/partition_locking.result
index 72ba78dd2de..9c41e99d035 100644
--- a/mysql-test/r/partition_locking.result
+++ b/mysql-test/r/partition_locking.result
@@ -5030,9 +5030,9 @@ Handler_read_next	12
 UNLOCK TABLES;
 EXPLAIN SELECT * FROM t2 WHERE a = sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t2	p1	const	PRIMARY	PRIMARY	4	const	#	100.00	NULL
+1	SIMPLE	t2	p0,p1,p2	eq_ref	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	/* select#1 */ select '1' AS `a`,'1' AS `b` from `test`.`t2` where true
+Note	1003	/* select#1 */ select `test`.`t2`.`a` AS `a`,`test`.`t2`.`b` AS `b` from `test`.`t2` where (`test`.`t2`.`a` = <cache>(`sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 SELECT * FROM t2 WHERE a = sf_a_from_t1b_d('1');
@@ -5086,9 +5086,9 @@ Handler_read_next	12
 UNLOCK TABLES;
 EXPLAIN SELECT * FROM t2 WHERE a = 7 + sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t2	p2	const	PRIMARY	PRIMARY	4	const	#	100.00	NULL
+1	SIMPLE	t2	p0,p1,p2	eq_ref	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	/* select#1 */ select '8' AS `a`,'8' AS `b` from `test`.`t2` where true
+Note	1003	/* select#1 */ select `test`.`t2`.`a` AS `a`,`test`.`t2`.`b` AS `b` from `test`.`t2` where (`test`.`t2`.`a` = <cache>((7 + `sf_a_from_t1b_d`('1'))))
 FLUSH STATUS;
 START TRANSACTION;
 SELECT * FROM t2 WHERE a = 7 + sf_a_from_t1b_d('1');
@@ -5138,9 +5138,9 @@ Handler_read_next	7
 UNLOCK TABLES;
 EXPLAIN SELECT * FROM t2 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	NULL	NULL	NULL	NULL	NULL	NULL	NULL	#	NULL	Impossible WHERE
+1	SIMPLE	t2	p2	const	PRIMARY	PRIMARY	4	const	#	100.00	NULL
 Warnings:
-Note	1003	/* select#1 */ select `test`.`t2`.`a` AS `a`,`test`.`t2`.`b` AS `b` from `test`.`t2` where false
+Note	1003	/* select#1 */ select '2' AS `a`,'2' AS `b` from `test`.`t2` where ((2 = `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 SELECT * FROM t2 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
@@ -5322,9 +5322,9 @@ Handler_read_next	12
 UNLOCK TABLES;
 EXPLAIN SELECT sf_add_1(a) - 1, sf_add_hello(b) FROM t2 WHERE a = sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t2	p1	const	PRIMARY	PRIMARY	4	const	#	100.00	NULL
+1	SIMPLE	t2	p0,p1,p2	eq_ref	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	/* select#1 */ select (`sf_add_1`('1') - 1) AS `sf_add_1(a) - 1`,`sf_add_hello`('1') AS `sf_add_hello(b)` from `test`.`t2` where true
+Note	1003	/* select#1 */ select (`sf_add_1`(`test`.`t2`.`a`) - 1) AS `sf_add_1(a) - 1`,`sf_add_hello`(`test`.`t2`.`b`) AS `sf_add_hello(b)` from `test`.`t2` where (`test`.`t2`.`a` = <cache>(`sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 SELECT sf_add_1(a) - 1, sf_add_hello(b) FROM t2 WHERE a = sf_a_from_t1b_d('1');
@@ -5378,9 +5378,9 @@ Handler_read_next	12
 UNLOCK TABLES;
 EXPLAIN SELECT sf_add_1(a) - 1, sf_add_hello(b) FROM t2 WHERE a = 7 + sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t2	p2	const	PRIMARY	PRIMARY	4	const	#	100.00	NULL
+1	SIMPLE	t2	p0,p1,p2	eq_ref	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	/* select#1 */ select (`sf_add_1`('8') - 1) AS `sf_add_1(a) - 1`,`sf_add_hello`('8') AS `sf_add_hello(b)` from `test`.`t2` where true
+Note	1003	/* select#1 */ select (`sf_add_1`(`test`.`t2`.`a`) - 1) AS `sf_add_1(a) - 1`,`sf_add_hello`(`test`.`t2`.`b`) AS `sf_add_hello(b)` from `test`.`t2` where (`test`.`t2`.`a` = <cache>((7 + `sf_a_from_t1b_d`('1'))))
 FLUSH STATUS;
 START TRANSACTION;
 SELECT sf_add_1(a) - 1, sf_add_hello(b) FROM t2 WHERE a = 7 + sf_a_from_t1b_d('1');
@@ -5430,9 +5430,9 @@ Handler_read_next	7
 UNLOCK TABLES;
 EXPLAIN SELECT sf_add_1(a) - 1, sf_add_hello(b) FROM t2 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	NULL	NULL	NULL	NULL	NULL	NULL	NULL	#	NULL	Impossible WHERE
+1	SIMPLE	t2	p2	const	PRIMARY	PRIMARY	4	const	#	100.00	NULL
 Warnings:
-Note	1003	/* select#1 */ select (`sf_add_1`(`test`.`t2`.`a`) - 1) AS `sf_add_1(a) - 1`,`sf_add_hello`(`test`.`t2`.`b`) AS `sf_add_hello(b)` from `test`.`t2` where false
+Note	1003	/* select#1 */ select (`sf_add_1`('2') - 1) AS `sf_add_1(a) - 1`,`sf_add_hello`('2') AS `sf_add_hello(b)` from `test`.`t2` where ((2 = `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 SELECT sf_add_1(a) - 1, sf_add_hello(b) FROM t2 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
@@ -5652,7 +5652,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET b = CONCAT('+', b) WHERE a = sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	t2	p1	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	UPDATE	t2	p0,p1,p2	index	NULL	PRIMARY	4	NULL	#	100.00	Using where
 Warnings:
 Note	1003	update `test`.`t2` set `test`.`t2`.`b` = concat('+',`test`.`t2`.`b`) where (`test`.`t2`.`a` = `sf_a_from_t1b_d`('1'))
 FLUSH STATUS;
@@ -5734,7 +5734,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET b = CONCAT('+', b) WHERE a = 7 + sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	UPDATE	t2	p0,p1,p2	index	NULL	PRIMARY	4	NULL	#	100.00	Using where
 Warnings:
 Note	1003	update `test`.`t2` set `test`.`t2`.`b` = concat('+',`test`.`t2`.`b`) where (`test`.`t2`.`a` = (7 + `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
@@ -5808,9 +5808,9 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET b = CONCAT('+', b) WHERE a = sf_a_from_t1b_d('1') AND a = 2;
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	NULL	NULL	NULL	NULL	NULL	NULL	NULL	#	NULL	Impossible WHERE
+1	UPDATE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	update `test`.`t2` set `test`.`t2`.`b` = concat('+',`test`.`t2`.`b`) where ()
+Note	1003	update `test`.`t2` set `test`.`t2`.`b` = concat('+',`test`.`t2`.`b`) where ((`test`.`t2`.`a` = 2) and (2 = `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 UPDATE t2 SET b = CONCAT('+', b) WHERE a = sf_a_from_t1b_d('1') AND a = 2;
@@ -6064,7 +6064,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET b = sf_add_hello(b) WHERE a = sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	t2	p1	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	UPDATE	t2	p0,p1,p2	index	NULL	PRIMARY	4	NULL	#	100.00	Using where
 Warnings:
 Note	1003	update `test`.`t2` set `test`.`t2`.`b` = `sf_add_hello`(`test`.`t2`.`b`) where (`test`.`t2`.`a` = `sf_a_from_t1b_d`('1'))
 FLUSH STATUS;
@@ -6146,7 +6146,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET b = sf_add_hello(b) WHERE a = 7 + sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	UPDATE	t2	p0,p1,p2	index	NULL	PRIMARY	4	NULL	#	100.00	Using where
 Warnings:
 Note	1003	update `test`.`t2` set `test`.`t2`.`b` = `sf_add_hello`(`test`.`t2`.`b`) where (`test`.`t2`.`a` = (7 + `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
@@ -6220,9 +6220,9 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET b = sf_add_hello(b) WHERE a = sf_a_from_t1b_d('1') AND a = 2;
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	NULL	NULL	NULL	NULL	NULL	NULL	NULL	#	NULL	Impossible WHERE
+1	UPDATE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	update `test`.`t2` set `test`.`t2`.`b` = `sf_add_hello`(`test`.`t2`.`b`) where ()
+Note	1003	update `test`.`t2` set `test`.`t2`.`b` = `sf_add_hello`(`test`.`t2`.`b`) where ((`test`.`t2`.`a` = 2) and (2 = `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 UPDATE t2 SET b = sf_add_hello(b) WHERE a = sf_a_from_t1b_d('1') AND a = 2;
@@ -6482,7 +6482,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET a = sf_add_1(a) + 4 WHERE a = sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	t2	p1	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	UPDATE	t2	p0,p1,p2	index	NULL	PRIMARY	4	NULL	#	100.00	Using where; Using temporary
 Warnings:
 Note	1003	update `test`.`t2` set `test`.`t2`.`a` = (`sf_add_1`(`test`.`t2`.`a`) + 4) where (`test`.`t2`.`a` = `sf_a_from_t1b_d`('1'))
 FLUSH STATUS;
@@ -6568,7 +6568,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET a = sf_add_1(a) + 4 WHERE a = 7 + sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	UPDATE	t2	p0,p1,p2	index	NULL	PRIMARY	4	NULL	#	100.00	Using where; Using temporary
 Warnings:
 Note	1003	update `test`.`t2` set `test`.`t2`.`a` = (`sf_add_1`(`test`.`t2`.`a`) + 4) where (`test`.`t2`.`a` = (7 + `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
@@ -6644,9 +6644,9 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN UPDATE t2 SET a = sf_add_1(a) + 4 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	UPDATE	NULL	NULL	NULL	NULL	NULL	NULL	NULL	#	NULL	Impossible WHERE
+1	UPDATE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	update `test`.`t2` set `test`.`t2`.`a` = (`sf_add_1`(`test`.`t2`.`a`) + 4) where ()
+Note	1003	update `test`.`t2` set `test`.`t2`.`a` = (`sf_add_1`(`test`.`t2`.`a`) + 4) where ((`test`.`t2`.`a` = 2) and (2 = `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 UPDATE t2 SET a = sf_add_1(a) + 4 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
@@ -6899,7 +6899,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN DELETE FROM t2 WHERE a = sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	DELETE	t2	p1	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	DELETE	t2	p0,p1,p2	ALL	NULL	NULL	NULL	NULL	#	100.00	Using where
 Warnings:
 Note	1003	delete from `test`.`t2` where (`test`.`t2`.`a` = `sf_a_from_t1b_d`('1'))
 FLUSH STATUS;
@@ -6977,7 +6977,7 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN DELETE FROM t2 WHERE a = 7 + sf_a_from_t1b_d('1');
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	DELETE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
+1	DELETE	t2	p0,p1,p2	ALL	NULL	NULL	NULL	NULL	#	100.00	Using where
 Warnings:
 Note	1003	delete from `test`.`t2` where (`test`.`t2`.`a` = (7 + `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
@@ -7049,9 +7049,9 @@ ROLLBACK;
 UNLOCK TABLES;
 EXPLAIN DELETE FROM t2 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	DELETE	NULL	NULL	NULL	NULL	NULL	NULL	NULL	#	NULL	Impossible WHERE
+1	DELETE	t2	p2	range	PRIMARY	PRIMARY	4	const	#	100.00	Using where
 Warnings:
-Note	1003	delete from `test`.`t2` where ()
+Note	1003	delete from `test`.`t2` where ((`test`.`t2`.`a` = 2) and (2 = `sf_a_from_t1b_d`('1')))
 FLUSH STATUS;
 START TRANSACTION;
 DELETE FROM t2 WHERE a = sf_a_from_t1b_d('1') AND a = 2;
diff --git a/mysql-test/r/sp.result b/mysql-test/r/sp.result
index f6a4657f82d..a7e413ef53c 100644
--- a/mysql-test/r/sp.result
+++ b/mysql-test/r/sp.result
@@ -6377,9 +6377,9 @@ Warnings:
 Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = 1)
 EXPLAIN SELECT * FROM t1 WHERE c1=f1();
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t1	NULL	ref	c1	c1	5	const	1	100.00	Using index
+1	SIMPLE	t1	NULL	ref	c1	c1	5	const	1	100.00	Using where; Using index
 Warnings:
-Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = `f1`())
+Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = <cache>(`f1`()))
 EXPLAIN SELECT * FROM v1 WHERE c1=1;
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
 1	SIMPLE	t1	NULL	ref	c1	c1	5	const	1	100.00	Using index
@@ -6387,14 +6387,14 @@ Warnings:
 Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = 1)
 EXPLAIN SELECT * FROM v1 WHERE c1=f1();
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t1	NULL	ref	c1	c1	5	const	1	100.00	Using index
+1	SIMPLE	t1	NULL	ref	c1	c1	5	const	1	100.00	Using where; Using index
 Warnings:
-Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = `f1`())
+Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = <cache>(`f1`()))
 EXPLAIN SELECT * FROM t1 WHERE c1=f2(10);
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
-1	SIMPLE	t1	NULL	ref	c1	c1	5	const	1	100.00	Using index
+1	SIMPLE	t1	NULL	ref	c1	c1	5	const	1	100.00	Using where; Using index
 Warnings:
-Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = `f2`(10))
+Note	1003	/* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` = <cache>(`f2`(10)))
 EXPLAIN SELECT * FROM t1 WHERE c1=f2(c1);
 id	select_type	table	partitions	type	possible_keys	key	key_len	ref	rows	filtered	Extra
 1	SIMPLE	t1	NULL	index	NULL	c1	5	NULL	5	20.00	Using where; Using index
diff --git a/mysql-test/t/partition_locking.test b/mysql-test/t/partition_locking.test
index 7a13b4c226a..7d9d8a1c469 100644
--- a/mysql-test/t/partition_locking.test
+++ b/mysql-test/t/partition_locking.test
@@ -2938,6 +2938,11 @@ DROP TABLE t1, t2;
 --echo #
 --echo # Test subqueries/stored functions with UPDATE/DELETE/SELECT
 --echo #
+# Bug#37560280:
+# The EXPLAIN plan for queries with the sf_a_from_t1b_d() function call
+# differs from the actual execution plan because constant folding
+# is performed in the actual plan but not in EXPLAIN.
+
 CREATE TABLE tq (id int PRIMARY KEY auto_increment, query varchar(255), not_select tinyint);
 CREATE TABLE tsq (id int PRIMARY KEY auto_increment, subquery varchar(255), can_be_locked tinyint);
 CREATE TABLE t1 (a int, b varchar(255), PRIMARY KEY (a), KEY (b))
diff --git a/sql/item.h b/sql/item.h
index 56eb35b704f..a7f927cc685 100644
--- a/sql/item.h
+++ b/sql/item.h
@@ -3594,6 +3594,15 @@ inline bool WalkItem(Item *item, enum_walk walk, T &&functor) {
                     reinterpret_cast<uchar *>(&functor));
 }
 
+/**
+   Overload for const 'item' and functor taking 'const Item*' argument.
+*/
+template <class T>
+inline bool WalkItem(const Item *item, enum_walk walk, T &&functor) {
+  auto to_const = [&](const Item *descendant) { return functor(descendant); };
+  return WalkItem(const_cast<Item *>(item), walk, to_const);
+}
+
 /**
   Same as WalkItem, but for Item::compile(). Use as e.g.:
 
diff --git a/sql/item_func.cc b/sql/item_func.cc
index 845aafffa5c..1eeb28e4a32 100644
--- a/sql/item_func.cc
+++ b/sql/item_func.cc
@@ -8192,6 +8192,8 @@ bool Item_func_sp::val_json(Json_wrapper *result) {
 bool Item_func_sp::execute() {
   THD *thd = current_thd;
 
+  assert(!thd->lex->is_explain() || thd->lex->is_explain_analyze);
+
   Internal_error_handler_holder<View_error_handler, Table_ref> view_handler(
       thd, context->view_error_handler, context->view_error_handler_arg);
 
diff --git a/sql/sql_derived.cc b/sql/sql_derived.cc
index 040db3695fa..972c536c147 100644
--- a/sql/sql_derived.cc
+++ b/sql/sql_derived.cc
@@ -1623,7 +1623,7 @@ bool Table_ref::optimize_derived(THD *thd) {
   // at execution time (in fact, it will get confused and crash if it has
   // already been materialized).
   if (!thd->lex->using_hypergraph_optimizer()) {
-    if (materializable_is_const() &&
+    if (materializable_is_const(thd) &&
         (create_materialized_table(thd) || materialize_derived(thd)))
       return true;
   }
diff --git a/sql/sql_lex.h b/sql/sql_lex.h
index b2e39db8ddf..96ebc92870d 100644
--- a/sql/sql_lex.h
+++ b/sql/sql_lex.h
@@ -726,6 +726,8 @@ class Query_expression {
   bool prepared;   ///< All query blocks in query expression are prepared
   bool optimized;  ///< All query blocks in query expression are optimized
   bool executed;   ///< Query expression has been executed
+  ///< Explain mode: query expression refers stored function
+  bool m_has_stored_program{false};
 
   /// Object to which the result for this query expression is sent.
   /// Not used if we materialize directly into a parent query expression's
@@ -770,6 +772,8 @@ class Query_expression {
   /// multi-level ORDER, i.e. we have a "simple table".
   bool is_simple() const { return m_query_term->term_type() == QT_QUERY_BLOCK; }
 
+  bool has_stored_program() const { return m_has_stored_program; }
+
   /// Values for Query_expression::cleaned
   enum enum_clean_state {
     UC_DIRTY,       ///< Unit isn't cleaned
@@ -4825,4 +4829,12 @@ bool accept_for_join(mem_root_deque<Table_ref *> *tables,
 Table_ref *nest_join(THD *thd, Query_block *select, Table_ref *embedding,
                      mem_root_deque<Table_ref *> *jlist, size_t table_cnt,
                      const char *legend);
+
+template <typename T>
+inline bool WalkQueryExpression(Query_expression *query_expr, enum_walk walk,
+                                T &&functor) {
+  return query_expr->walk(&Item::walk_helper_thunk<T>, walk,
+                          reinterpret_cast<uchar *>(&functor));
+}
+
 #endif /* SQL_LEX_INCLUDED */
diff --git a/sql/sql_optimizer.cc b/sql/sql_optimizer.cc
index 7bdfdd37f41..01c6d848d12 100644
--- a/sql/sql_optimizer.cc
+++ b/sql/sql_optimizer.cc
@@ -5652,12 +5652,16 @@ bool JOIN::extract_const_tables() {
            1. are dependent upon other tables, or
            2. have no exact statistics, or
            3. are full-text searched
+           4. a derived table which has a stored function
         */
+        const bool explain_mode = thd->lex->is_explain();
         if ((table->s->system || table->file->stats.records <= 1 ||
              all_partitions_pruned_away) &&
             !tab->dependent &&                                              // 1
             (table->file->ha_table_flags() & HA_STATS_RECORDS_IS_EXACT) &&  // 2
-            !tl->is_fulltext_searched())                                    // 3
+            !tl->is_fulltext_searched() &&                                  // 3
+            !(explain_mode && tl->is_view_or_derived() &&
+              tl->has_stored_program()))  // 4
           mark_const_table(tab, nullptr);
         break;
     }
@@ -5811,6 +5815,7 @@ bool JOIN::extract_func_dependent_tables() {
              6. are not going to be used, typically because they are streamed
                 instead of materialized
                 (see Query_expression::can_materialize_directly_into_result()).
+             7. key evaluated in stored program in EXPLAIN mode
           */
           if (eq_part.is_prefix(table->key_info[key].user_defined_key_parts) &&
               !tl->is_fulltext_searched() &&                              // 1
@@ -5818,7 +5823,9 @@ bool JOIN::extract_func_dependent_tables() {
               !(tl->embedding && tl->embedding->is_sj_or_aj_nest()) &&    // 3
               !(tab->join_cond() && tab->join_cond()->is_expensive()) &&  // 4
               !(table->file->ha_table_flags() & HA_BLOCK_CONST_TABLE) &&  // 5
-              table->is_created()) {                                      // 6
+              table->is_created() &&                                      // 6
+              !(thd->lex->is_explain() &&
+                start_keyuse->val->has_stored_program())) {  // 7
             if (table->key_info[key].flags & HA_NOSAME) {
               if (const_ref == eq_part) {  // Found everything for ref.
                 ref_changed = true;
@@ -6249,7 +6256,7 @@ static ha_rows get_quick_record_count(THD *thd, JOIN_TAB *tab, ha_rows limit,
       return 0;
     }
     DBUG_PRINT("warning", ("Couldn't use record count on const keypart"));
-  } else if (tl->is_table_function() || tl->materializable_is_const()) {
+  } else if (tl->is_table_function() || tl->materializable_is_const(thd)) {
     tl->fetch_number_of_rows();
     return tl->table->file->stats.records;
   }
@@ -9005,9 +9012,11 @@ static bool test_if_ref(THD *thd, Item_field *left_item, Item *right_item,
       (join_tab->type() != JT_REF_OR_NULL)) {
     Item *ref_item = part_of_refkey(field->table, &join_tab->ref(), field);
     if (ref_item != nullptr && ref_item->eq(right_item, true)) {
-      if (ref_lookup_subsumes_comparison(thd, field, right_item,
-                                         right_item->const_for_execution(),
-                                         redundant)) {
+      if (ref_lookup_subsumes_comparison(
+              thd, field, right_item,
+              right_item->const_for_execution() &&
+                  !(thd->lex->is_explain() && right_item->has_stored_program()),
+              redundant)) {
         return true;
       }
     }
@@ -11456,6 +11465,13 @@ bool evaluate_during_optimization(const Item *item, const Query_block *select) {
   // If the Item does not access any tables, it can always be evaluated.
   if (item->const_item()) return true;
 
+  // Do not evaluate stored procedure in EXPLAIN
+  if (current_thd->lex->is_explain() &&
+      WalkItem(item, enum_walk::PREFIX, [](const Item *curitem) {
+        return curitem->has_stored_program();
+      }))
+    return false;
+
   return !item->has_subquery() || (select->active_options() &
                                    OPTION_NO_SUBQUERY_DURING_OPTIMIZATION) == 0;
 }
diff --git a/sql/sql_select.cc b/sql/sql_select.cc
index 0d38a781d17..c86ff60d7bd 100644
--- a/sql/sql_select.cc
+++ b/sql/sql_select.cc
@@ -2334,8 +2334,8 @@ bool init_ref_part(THD *thd, unsigned part_no, Item *val, bool *cond_guard,
                                    key_part_info, key_buff, nullable);
   if (unlikely(!s_key || thd->is_error())) return true;
 
-  if (used_tables & ~INNER_TABLE_BIT) {
-    /* Comparing against a non-constant. */
+  if (used_tables & ~INNER_TABLE_BIT ||
+      (thd->lex->is_explain() && val->has_stored_program())) {
     ref->key_copy[part_no] = s_key;
   } else {
     /*
diff --git a/sql/sql_union.cc b/sql/sql_union.cc
index 8be63366c2b..7c7e6f7d49b 100644
--- a/sql/sql_union.cc
+++ b/sql/sql_union.cc
@@ -866,6 +866,18 @@ bool Query_expression::prepare(THD *thd, Query_result *sel_result,
     for (size_t i = 0; i < types.size(); i++)
       types[i]->set_nullable(nullable[i]);
   }
+  if (thd->lex->is_explain()) {
+    WalkQueryExpression(this,
+                        enum_walk::SUBQUERY_POSTFIX,  // Use SUBQUERY_POSTFIX to
+                                                      // traverse subqueries
+                        [this](Item *item) {
+                          if (item->has_stored_program()) {
+                            this->m_has_stored_program = true;
+                            return true;  // Stop walking
+                          }
+                          return false;  // Continue walking
+                        });
+  }
 
   // Query blocks are prepared, update the state
   set_prepared();
diff --git a/sql/table.cc b/sql/table.cc
index c34f519716a..35ae4e79f63 100644
--- a/sql/table.cc
+++ b/sql/table.cc
@@ -6498,12 +6498,19 @@ bool Table_ref::is_mergeable() const {
   return derived->is_mergeable();
 }
 
-bool Table_ref::materializable_is_const() const {
+bool Table_ref::has_stored_program() const {
+  assert(derived != nullptr);
+  return derived->has_stored_program();
+}
+
+bool Table_ref::materializable_is_const(THD *thd) const {
   assert(uses_materialization());
   const Query_expression *unit = derived_query_expression();
+  const bool explain_mode = thd->lex->is_explain();
   return unit->query_result()->estimated_rowcount <= 1 &&
          (unit->first_query_block()->active_options() &
-          OPTION_NO_SUBQUERY_DURING_OPTIMIZATION) == 0;
+          OPTION_NO_SUBQUERY_DURING_OPTIMIZATION) == 0 &&
+         !(explain_mode && has_stored_program());
 }
 
 /**
diff --git a/sql/table.h b/sql/table.h
index cd90d6d82d3..9e467a8ffb6 100644
--- a/sql/table.h
+++ b/sql/table.h
@@ -3118,7 +3118,10 @@ class Table_ref {
     during execution. The hypergraph optimizer does not care about const tables,
     so such tables are not executed during optimization time when it is active.
   */
-  bool materializable_is_const() const;
+  bool materializable_is_const(THD *thd) const;
+
+  /// @returns true if this is a derived table containing a stored function.
+  bool has_stored_program() const;
 
   /// Return true if this is a derived table or view that is merged
   bool is_merged() const { return effective_algorithm == VIEW_ALGORITHM_MERGE; }
